### PR TITLE
workflows: Changes to allow debugging "source forest creation failed"

### PR DESCRIPTION
**Pull request by bduffany**
* Add a CI runner image which uses a debug build of bazel that prints more information whenever "Source forest creation failed" happens.
* Allow configuring `--remote_execution.workflows_default_image` to override the CI runner image. (Named it "default" image to future-proof in case we want to allow people to set it via buildbuddy.yaml)
* Allow configuring `--remote_execution.workflows_ci_runner_debug` which causes the CI runner to run `ls -la` before and after each action, so that we can see the permissions on each file in the workspace and see whether they get into a state that might cause the "Source forest creation failed" error.

---

**Version bump**: Patch

### CODE REVIEW

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/cmd/ci_runner/main.go [hunk 1]
```diff
@@ -90,10 +91,17 @@ var (
...
+
 	// Test-only flags
 	fallbackToCleanCheckout = flag.Bool("fallback_to_clean_checkout", true, "Fallback to cloning the repo from scratch if sync fails (for testing purposes only).")
 
 	shellCharsRequiringQuote = regexp.MustCompile(`[^\w@%+=:,./-]`)
+
+	// initLogs contain informational logs from the setup phase (cloning the
+	// git repo and deciding which actions to run) which are reported as part of
+	// the first action's logs.
+	initLogs bytes.Buffer
```
[the remark this answers is not in the document]
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go]: Done

File: enterprise/server/cmd/ci_runner/main.go [hunk 2]
```diff
@@ -423,7 +433,32 @@ type actionRunner struct {
 	progressCount int32
 }
 
+func initLog(f string, args ...interface{}) {
+	initLogs.Write([]byte(fmt.Sprintf(f+"\n", args...)))
+}
```
[the remark this answers is not in the document]
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go]: Oh nice, TIL about WriteString. Done

File: enterprise/server/cmd/ci_runner/main.go [hunk 3]
```diff
@@ -423,7 +433,32 @@ type actionRunner struct {
...
+
+func runDebugScript(out io.Writer, script string) {
```
[the remark this answers is not in the document]
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go:436]: Done

File: enterprise/server/workflow/service/service.go [hunk 4]
```diff
@@ -41,10 +41,14 @@ import (
...
+	workflowsImage = "docker://gcr.io/flame-public/buildbuddy-ci-runner:v1.7.1"
+)
+
 var (
-	workflowURLMatcher   = regexp.MustCompile(`^.*/webhooks/workflow/(?P<instance_name>.*)$`)
-	buildbuddyCIUserName = "buildbuddy"
-	buildbuddyCIHostName = "buildbuddy-ci-runner"
+	ciRunnerDebug = flag.Bool("ci_runner_debug", false, "Pass --debug to the CI runner.")
+
+	workflowURLMatcher = regexp.MustCompile(`^.*/webhooks/workflow/(?P<instance_name>.*)$`)
```
[the remark this answers is not in the document]
Comment (bduffany) [reply enterprise/server/workflow/service/service.go]: Done

File: enterprise/server/cmd/ci_runner/main.go [hunk 5]
```diff
@@ -423,7 +433,26 @@ type actionRunner struct {
...
+		return
+	}
+	out.Write([]byte(string(fmt.Sprintf("(debug) # %s\n", script))))
+	output, err := exec.Command("sh", "-c", script).CombinedOutput()
+	out.Write(output)
+	exitCode := getExitCode(err)
+	if exitCode != noExitCode {
+		out.Write([]byte(fmt.Sprintf("%s(command exited with code %d)%s\n", ansiGray, exitCode, ansiReset)))
+	}
+	out.Write([]byte("===\n"))
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:447]: WriteString here and here

File: enterprise/server/cmd/ci_runner/main.go [hunk 6]
```diff
@@ -423,7 +433,26 @@ type actionRunner struct {
 	progressCount int32
 }
 
+func runDebugScript(out io.Writer, script string) {
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:436]: script is a slightly odd name for this since you use it to run commands?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
LGTM (some nits)

File: enterprise/server/cmd/ci_runner/main.go (hunk 6 above)
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go]: Renamed to runDebugCommand

File: enterprise/server/cmd/ci_runner/main.go [hunk 7]
```diff
@@ -423,7 +433,26 @@ type actionRunner struct {
...
+	if !*debug {
+		return
+	}
+	out.Write([]byte(string(fmt.Sprintf("(debug) # %s\n", script))))
```
[the remark this answers is not in the document]
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go]: Done

File: enterprise/server/cmd/ci_runner/main.go (hunk 5 above)
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go]: Done

**Review by tylerwilliams — APPROVED** [review]
LGTM

**Pull request merged**